### PR TITLE
fix: update mkldnn version to v0.17 issues.

### PR DESCRIPTION
**Pull request by i8run**
## What changes were proposed in this pull request?
Upgrade mkldnn to v0.17.

There're two issues,

1. the padding tensor required. mkl-dnn will use a padding tensor which
   will use more memory, such as 4x1x28x28 to 4x8x28x28(avx2). It will
   pad to times of simd width.
2. the TensorMMap between DenseTensor and DnnTensor. Previous impl
    will allocate DnnTensor when model is created, which will cost too much
    space. So this patch will allocate it at runtime.

## How was this patch tested?

Test on ResNet-50 training. Test on Jenkins.

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala [hunk 1]
```diff
@@ -84,6 +85,11 @@ class ReorderMemory(inputFormat: MemoryData, outputFormat: MemoryData,
 
     // recover to original data
     output = initTensor(realOutput(0))
+
+    if (outputLayout == Memory.Format.nhwc && realOutput(0).isInstanceOf[HeapData]) {
+      output.toTensor[Float].resize(_outputFormats(0).shape)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: if not heapdata, output may not have real size, is it right?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala [hunk 2]
```diff
@@ -85,6 +94,10 @@ class ThreadPool(private var poolSize: Int) {
       val tid = Thread.currentThread().getId()
       logger.info(s"Set mkl threads to $size on thread $tid")
     }(context)).foreach(Await.result(_, Duration.Inf))
+
+    invokeAndWait2 ((0 until poolSize).map( _ => () => {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: Why do you need this?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala (hunk 1 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/ReorderMemory.scala]: The tensor size will always be right.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala [hunk 3]
```diff
@@ -138,8 +136,8 @@ class Linear(
     updateWithNewTensor(updateOutputTensors, 0, input)
 
     if (isTraining()) {
-      weight.syncToNative()
-      bias.syncToNative()
+      weight.sync()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala]: if not training, when to sync weight and bias to native?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/SpatialConvolution.scala [hunk 4]
```diff
@@ -253,8 +250,8 @@ class SpatialConvolution(
     updateWithNewTensor(updateOutputTensors, 0, input)
 
     if (isTraining()) {
-      weight.syncToNative()
-      bias.syncToNative()
+      weight.sync()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/SpatialConvolution.scala]: if not training, when to sync?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala [hunk 5]
```diff
@@ -58,26 +58,24 @@ class Linear(
...
   }
 
   override private[mkldnn] def initFwdPrimitives(inputs: Array[MemoryData], phase: Phase) = {
-    val weightShape = inputs(0).shape.length match {
-      case 4 => Array(weight.size(1)) ++ inputs(0).shape.slice(1, 4)
-      case _ => weight.size()
+    val (weightShape, weightLayout) = inputs(0).shape.length match {
+      case 4 =>
+        (Array(weight.size(1)) ++ inputs(0).shape.slice(1, 4),
+          Memory.Format.oihw)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala]: what's oihw for

Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala]: Linear (mkldnn version) will support 4-D input, so the relevant weight will be 4-D. We make all 4-D weight with the format oihw.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala (hunk 3 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Linear.scala]: the previous commit will sync weight and bias at `setMemoryData`. Have pushed a new commit which will sync weight hard code when compiling.

**Comment by i8run** [comment]
Jenkins has passed all test cases. And distributed training of ResNet-50 is right.

**Pull request merged**